// @google-cloud/aiplatform/build/src/v1/index.js
"use strict";
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by gapic-generator-typescript. **
// ** https://github.com/googleapis/gapic-generator-typescript **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpecialistPoolServiceClient = exports.PredictionServiceClient = exports.PipelineServiceClient = exports.ModelServiceClient = exports.MigrationServiceClient = exports.JobServiceClient = exports.EndpointServiceClient = exports.DatasetServiceClient = void 0;
var dataset_service_client_1 = require("./dataset_service_client");
Object.defineProperty(exports, "DatasetServiceClient", { enumerable: true, get: function () { return dataset_service_client_1.DatasetServiceClient; } });
var endpoint_service_client_1 = require("./endpoint_service_client");
Object.defineProperty(exports, "EndpointServiceClient", { enumerable: true, get: function () { return endpoint_service_client_1.EndpointServiceClient; } });
var job_service_client_1 = require("./job_service_client");
Object.defineProperty(exports, "JobServiceClient", { enumerable: true, get: function () { return job_service_client_1.JobServiceClient; } });
var migration_service_client_1 = require("./migration_service_client");
Object.defineProperty(exports, "MigrationServiceClient", { enumerable: true, get: function () { return migration_service_client_1.MigrationServiceClient; } });
var model_service_client_1 = require("./model_service_client");
Object.defineProperty(exports, "ModelServiceClient", { enumerable: true, get: function () { return model_service_client_1.ModelServiceClient; } });
var pipeline_service_client_1 = require("./pipeline_service_client");
Object.defineProperty(exports, "PipelineServiceClient", { enumerable: true, get: function () { return pipeline_service_client_1.PipelineServiceClient; } });
var prediction_service_client_1 = require("./prediction_service_client");
Object.defineProperty(exports, "PredictionServiceClient", { enumerable: true, get: function () { return prediction_service_client_1.PredictionServiceClient; } });
var specialist_pool_service_client_1 = require("./specialist_pool_service_client");
Object.defineProperty(exports, "SpecialistPoolServiceClient", { enumerable: true, get: function () { return specialist_pool_service_client_1.SpecialistPoolServiceClient; } });
