// @google-cloud/aiplatform/build/src/index.js
"use strict";
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by synthtool. **
// ** https://github.com/googleapis/synthtool **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.helpers = exports.protos = exports.SpecialistPoolServiceClient = exports.PredictionServiceClient = exports.PipelineServiceClient = exports.ModelServiceClient = exports.MigrationServiceClient = exports.JobServiceClient = exports.EndpointServiceClient = exports.DatasetServiceClient = exports.v1 = exports.v1beta1 = void 0;
const v1beta1 = require("./v1beta1");
exports.v1beta1 = v1beta1;
const v1 = require("./v1");
exports.v1 = v1;
const DatasetServiceClient = v1.DatasetServiceClient;
exports.DatasetServiceClient = DatasetServiceClient;
const EndpointServiceClient = v1.EndpointServiceClient;
exports.EndpointServiceClient = EndpointServiceClient;
const JobServiceClient = v1.JobServiceClient;
exports.JobServiceClient = JobServiceClient;
const MigrationServiceClient = v1.MigrationServiceClient;
exports.MigrationServiceClient = MigrationServiceClient;
const ModelServiceClient = v1.ModelServiceClient;
exports.ModelServiceClient = ModelServiceClient;
const PipelineServiceClient = v1.PipelineServiceClient;
exports.PipelineServiceClient = PipelineServiceClient;
const PredictionServiceClient = v1.PredictionServiceClient;
exports.PredictionServiceClient = PredictionServiceClient;
const SpecialistPoolServiceClient = v1.SpecialistPoolServiceClient;
exports.SpecialistPoolServiceClient = SpecialistPoolServiceClient;
exports.default = {
    v1beta1,
    v1,
    DatasetServiceClient,
    EndpointServiceClient,
    JobServiceClient,
    MigrationServiceClient,
    ModelServiceClient,
    PipelineServiceClient,
    PredictionServiceClient,
    SpecialistPoolServiceClient,
};
const protos = require("../protos/protos");
exports.protos = protos;
const helpers_1 = require("./helpers");
const helpers = { toValue: helpers_1.toValue, fromValue: helpers_1.fromValue };
exports.helpers = helpers;
const decorator_1 = require("./decorator");
decorator_1._enhance('v1beta1');
decorator_1._enhance('v1');
